src/wp-includes/class-wp-theme-json-resolver.php

```diff
@@ -243,7 +243,9 @@ public static function get_theme_data( $deprecated = array(), $options = array()
 			_deprecated_argument( __METHOD__, '5.9.0' );
 		}
 
-		$options = wp_parse_args( $options, array( 'with_supports' => true ) );
+		if ( static::$theme ) {
+			return static::$theme;
+		}
```

Comment:
This wouldn't work reliably, as we need to consider the value of `$options['with_supports']` every time this method is called.

How about we introduce a new class property here `$theme_with_supports` so that we can cache both values individually? That way, if `$options['with_supports']` is `true`, we can return `$theme_with_supports`, otherwise we return `$theme`.

Also, we must only return a cached value when the registered blocks haven't changed, so we still need to call `static::has_same_registered_blocks( 'theme' )` here.